### ISSUE TITLE
Packing messages into packets

### ISSUE THREAD

**Issue by cBournhonesque** [mas-bandwidth/yojimbo#200]
Hi!
I was wondering what your opinion was on how to pack messages into packets.

- Is it better to have one big message that I manually split between sub-messages of size MTU? (to avoid having the packet be fragmented, which makes it much more unreliable if there is any packet loss or jitter)
- Is it better to have lots of small messages? Then there might be additional overhead of having many `MessageId`s and other header information added.
- If I want to guarantee that 2 pieces of data are received together (for example I have data about a parent object and a child object. Because of these hierarchy I cannot have the data about the child arrive before the data about the parent arrives first), I need to put these 2 pieces of data into a single message?

**Comment by gafferongames** [comment]
Yojimbo lets you attach blocks of memory on messages. 

a) In the reliable channel, these blocks will be sliced up and sent reliably in packets smaller than MTU automatically, and the reliable channel will block any messages to be received after the message with block attach so they arrive after the block send completes, so messages will always be received reliably and in-order. 

b) Alternatively, if you need to send a large block of data and it's not too large, eg. 4-8 packets @ MTU worth, you could send it as a big fat unreliable message, and the system will fragment into MTU packets and re-assemble automatically for you, but if any of the fragments are dropped, the whole message is lost.

Basically, a) was designed for you to send down a large block of data on connect, or on load level or whatever on the reliable channel. b) was designed so that when you are sending delta encoded snapshots, it would allow you to go over MTU and it would do fragmentation and reassembly for you automatically, since some routers on the internet drop any packet fragments done at the IP level.

cheers

**Comment by gafferongames** [comment]
And to answer your question -- I need more information about what you are trying to do. The correct answer depends very much on whether you are using the reliable or unreliable channels to send messages, and what you are trying to achieve.

**Comment by cBournhonesque** [comment]
My question was not directly related to Yojimbo, sorry about that. (maybe I could send you an email directly if that is more appropriate channel)?
I've been developing a networking library for an ECS game engine as a hobby project: https://github.com/cBournhonesque/lightyear 
I've read all your articles many times as they provide a great starting point.

One thing i've been struggling is any guidance on how messages should fit into packets.
I have two types of messages: Actions (Entity spawn/Entity despawn/Component insertion/Component removal) and Updates (Component update). Actions are sent reliably, and Updates are sent unreliably.
